api/src/download.py

```diff
@@ -0,0 +1,106 @@
+import MySQLdb
+import csv
+
+from flask import Blueprint
+from flask import request as req
+from flask import abort
+from flask import send_file
+
+from datetime import datetime
+
+from schema import schema_col_names
+from config import load_config
+
+download_bp = Blueprint('download', __name__)
+
+config = load_config()
+
+@download_bp.route('/')
+def download():
+
+    # START : same as ballots.py endpoint from here to "END"
+
+    # required parameters - throws error if not present
+    try:
+        state = req.args['state'].upper()
+        elec_dt = datetime.strptime(req.args['election_dt'], '%m-%d-%Y')
+    except:
+        abort(404, description="Resource not found")
+    
+    # build WHERE clause for optional parameters on the fly for optimized SQL query times
+    where_clause = ''
+
+    # set any default values for params if needed
+    rtn_status = req.args.get('ballot_rtn_status', 'R')
+    where_clause += f'ballot_rtn_status = "{rtn_status}" AND '
+
+    # optional parameters
+    for param in schema_col_names:
+        if param in where_clause or param == 'state' or param == 'election_dt': 
+            continue
+        else:
+            val = req.args.get(param, None)
+            where_clause += f'{param} = "{val}" AND ' if val else '' # double quotes important to prevent SQL injection
+
+    # remove last AND
+    where_clause = where_clause[:-5]
+
+
+    limit = int(req.args.get('limit', 10))
+    limit_clause = f'LIMIT {limit}' if limit != -1 else ''
+
+    # TODO support historic data requests - run query on `rejected` table, otherwise run on main table
+    historic = req.args.get('show_historic', False)
+
+    try:
+        mydb = MySQLdb.connect(host=config['DATABASE']['host'],
+        user=config['DATABASE']['user'],
+        passwd=config['DATABASE']['passwd'],
+        db=config[state]['db'],
+        local_infile = 1)
+    except:
+        # if connection failed, then input state was not valid
+        abort(500, description="internal service failure")
+    
+    cur = mydb.cursor()
+
+    db_table_name = elec_dt.strftime('%m_%d_%Y')
+
+    query = f'''
+    SELECT *
+    FROM {db_table_name}
+    WHERE {where_clause}
+    {limit_clause};
+    '''
+
+    print(f'DEBUG:\n{query}')
+
+    try:
+        cur.execute(query)
+    except:
+        # if valid, then election_dt not valid
+        abort(500, description="internal service failure")
+
+    # get the row headers
+    row_headers = [x[0] for x in cur.description]
+    id_idx = row_headers.index('id')
+    row_headers.pop(id_idx)    
+
+    # END : end of similarity with ballots.py endpoint
+
+    # get the file name
+    filename = f"vote_{datetime.now().strftime('%d_%m_%y_%H_%M_%S')}.csv"
```

Comment:
I think it would be better to have the filename be consistent with our other naming schemas in our database and API. i.e. `%m_%d_%Y'. 

Since we only ingest once per day, the exact time isn't needed imo. If people want to keep track of that, it is still recorded in the metadata when they download the file.

Reply:
My reasoning behind this was that I was unsure what would happen if two users requested a file at nearly the same time such that both requests were being served at the same time. I wasn't sure if somehow if the file names were the same would one request potentially overwrite the file generated by the first request before being able to send it. Is this a case that would just be handled automatically by flask/server such that we wouldn't have to worry about it? Or is this just a such a niche case that it isn't worth worrying about?

Reply:
Actually, your intuitive reasoning is pretty good. I would keep it this way in the backend, but rename `attachment_filename` in the `send_file()` call so that the user who downloaded it will receive it in a consistent naming schema. By doing this, we also avoid large downloads not clogging the pipeline (as described below).



To give some answers to your questions though:
I think it might be a little niche, but if it's a large download it still might take a while to write to the file and send it. What happens here is that the request that gets to writing the file first puts a lock on it, so the second request would have to wait for the first request to release its file lock before it can start writing its own data. It _might_ be possible that the second request starts file writing before the first reaches its `send_file()` call, but my understanding is that the `send_file()` call will send the file before any changes from the second request would be applied, since those changes are still in the buffer and the file lock is still not released. But you can see that this area is sort of unsure of which request will reach what first, and we call these sorts of things data races.

If you are interested in explicitly handling data races (and this comes up in software engineering backend stuff all the time), this is where synchronization primitives [[1]](https://www.cs.columbia.edu/~hgs/os/sync.html) [[2]](https://docs.python.org/3/library/asyncio-sync.html) come into play. You can create your own lock (mutex) on the file and then release the lock after you send the file, so that the second request wouldn't even be able to start writing to the file until after the first one sends it. [This SO post](https://stackoverflow.com/questions/6931342/system-wide-mutex-in-python-on-linux) shows how to use a Python wrapper around a Unix file lock).  

Can look into/take OS if you're interested in this kind of stuff

Reply:
Ah thanks...I remembered some stuff about this from CSF but wasn't sure exactly how/if it would apply in this case. That makes sense though and is interesting. 

By the way, I updated the attachment_filename in send_file() as requested.